src/core/DY.Core.Bytes.fst

```diff
@@ -1916,3 +1990,232 @@ val get_label_kdf_expand:
+    get_kem_label (kem_pk sk) == get_label sk
+  ))
+  [SMTPat (get_kem_label (kem_pk sk))]
+let get_kem_label_kem_pk #cu sk =
+  normalize_term_spec get_kem_label;
+  normalize_term_spec kem_pk
+
+/// Lemma for attacker knowledge theorem.
+
+val kem_encap_preserves_publishability:
+  {|crypto_invariants|} -> tr:trace ->
+  pk:bytes -> nonce:bytes ->
+  Lemma
+  (requires
+    is_publishable tr pk /\
+    is_publishable tr nonce
+  )
+  (ensures (
+    let (kem_output, ss) = kem_encap pk nonce in
+    is_publishable tr kem_output /\
+    is_publishable tr ss
+  ))
+let kem_encap_preserves_publishability #ci tr pk nonce =
+  normalize_term_spec kem_encap;
+  normalize_term_spec bytes_invariant;
+  normalize_term_spec get_label;
+  assert(is_publishable tr (KemSecretShared nonce))
+
+/// Lemma for attacker knowledge theorem.
+
+val kem_decap_preserves_publishability:
+  {|crypto_invariants|} -> tr:trace ->
+  sk:bytes -> encap:bytes ->
+  Lemma
+  (requires
+    is_publishable tr sk /\
+    is_publishable tr encap
+  )
+  (ensures (
+    match kem_decap sk encap with
+    | Some ss -> is_publishable tr ss
+    | None -> True
+  ))
+let kem_decap_preserves_publishability #ci tr sk encap =
+  normalize_term_spec kem_decap;
+  normalize_term_spec bytes_invariant;
+  normalize_term_spec get_label;
+  normalize_term_spec get_kem_label;
+  match encap with
+  | KemEncap (KemPub sk') nonce ->
+    if sk = sk' then ()
+    else ()
+  | _ -> ()
+
+/// User lemma (kem_encap properties)
+
+val bytes_invariant_kem_encap:
+  {|crypto_invariants|} -> tr:trace ->
+  pk:bytes -> nonce:bytes ->
+  Lemma
+  (requires
+    is_publishable tr pk /\
+    bytes_invariant tr nonce /\
+    KemNonce? (get_usage nonce) /\
+    (get_label nonce) `can_flow tr` (get_kem_label pk) /\ (
+      get_label nonce `can_flow tr` public \/
+      get_kem_usage pk == KemKey (KemNonce?.usg (get_usage nonce))
+    )
+  )
+  (ensures (
+    let (kem_output, ss) = kem_encap pk nonce in
+    is_publishable tr kem_output /\
+    bytes_invariant tr ss /\
+    get_label ss == get_label nonce /\
+    get_usage ss == (KemNonce?.usg (get_usage nonce))
+  ))
+  [SMTPat (kem_encap pk nonce); SMTPat (bytes_invariant tr nonce)]
+let bytes_invariant_kem_encap #ci tr pk nonce =
+  normalize_term_spec kem_encap;
+  normalize_term_spec bytes_invariant;
+  normalize_term_spec get_usage;
+  normalize_term_spec get_kem_usage;
+  normalize_term_spec get_label;
+  normalize_term_spec get_kem_label;
+  let (kem_output, ss) = kem_encap pk nonce in
+  ()
+
+/// User lemma (kem_decap usage)
+
+#push-options "--z3rlimit 25"
+val get_usage_kem_decap:
+  {|crypto_invariants|} -> tr:trace ->
+  sk:bytes -> encap:bytes ->
+  Lemma
+  (requires
+    bytes_invariant tr sk /\
+    bytes_invariant tr encap /\
+    (KemKey? (get_usage sk) \/ (get_label sk) `can_flow tr` public)
+  )
+  (ensures (
+    match kem_decap sk encap with
+    | Some ss -> get_label ss `can_flow tr` public \/ get_usage sk == KemKey (get_usage ss)
```

Comment:
This is only a small thing, but in my opinion, the ensures would be better readable if we invert the order of the OR condition so that it says either it has this usage or it can flow to public.

Reply:
Done